DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.cpp

```diff
@@ -0,0 +1,185 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "ClientConnection.h"
+#include "AutoComplete.h"
+#include <AK/HashMap.h>
+#include <LibCore/File.h>
+#include <LibGUI/TextDocument.h>
+
+// #define DEBUG_CPP_LANGUAGE_SERVER
+// #define DEBUG_FILE_CONTENT
+
+namespace LanguageServers::Cpp {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<CppLanguageClientEndpoint, CppLanguageServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    exit(0);
+}
+
+OwnPtr<Messages::CppLanguageServer::GreetResponse> ClientConnection::handle(const Messages::CppLanguageServer::Greet& message)
+{
+    m_project_root = LexicalPath(message.project_root());
+#ifdef DEBUG_CPP_LANGUAGE_SERVER
+    dbg() << "project_root: " << m_project_root.string();
+#endif
+    return make<Messages::CppLanguageServer::GreetResponse>(client_id());
+}
+
+class DefaultDocumentClient final : public GUI::TextDocument::Client {
+public:
+    virtual ~DefaultDocumentClient() override = default;
+    virtual void document_did_append_line() override {};
+    virtual void document_did_insert_line(size_t) override {};
+    virtual void document_did_remove_line(size_t) override {};
+    virtual void document_did_remove_all_lines() override {};
+    virtual void document_did_change() override {};
+    virtual void document_did_set_text() override {};
+    virtual void document_did_set_cursor(const GUI::TextPosition&) override {};
+
+    virtual bool is_automatic_indentation_enabled() const override { return true; }
+    virtual int soft_tab_width() const { return 4; }
+};
+
+static DefaultDocumentClient s_default_document_client;
+
+void ClientConnection::handle(const Messages::CppLanguageServer::FileOpened& message)
+{
+    LexicalPath file_path(String::format("%s/%s", m_project_root.string().characters(), message.file_name().characters()));
+#ifdef DEBUG_CPP_LANGUAGE_SERVER
+    dbg() << "FileOpened: " << file_path.string();
+#endif
+
+    auto file = Core::File::construct(file_path.string());
+    if (!file->open(Core::IODevice::ReadOnly)) {
+        perror("open");
```

Comment:
Does `Core::File::open()` set `errno`? I thought it saved the error to a member int that you had to get explicitly...